packages/integration/src/ScmIntegrations.ts

```diff
@@ -21,27 +21,64 @@ import { GitHubIntegration } from './github/GitHubIntegration';
 import { GitLabIntegration } from './gitlab/GitLabIntegration';
 import {
   ScmIntegration,
-  ScmIntegrationPredicateTuple,
   ScmIntegrationRegistry,
+  ScmIntegrationsGroup,
 } from './types';
 
+type IntegrationsByType = {
+  azure: ScmIntegrationsGroup<AzureIntegration>;
+  bitbucket: ScmIntegrationsGroup<BitbucketIntegration>;
+  github: ScmIntegrationsGroup<GitHubIntegration>;
+  gitlab: ScmIntegrationsGroup<GitLabIntegration>;
+};
+
 export class ScmIntegrations implements ScmIntegrationRegistry {
+  private readonly byType: IntegrationsByType;
+
   static fromConfig(config: Config): ScmIntegrations {
-    return new ScmIntegrations([
-      ...AzureIntegration.factory({ config }),
-      ...BitbucketIntegration.factory({ config }),
-      ...GitHubIntegration.factory({ config }),
-      ...GitLabIntegration.factory({ config }),
-    ]);
+    return new ScmIntegrations({
+      azure: AzureIntegration.factory({ config }),
+      bitbucket: BitbucketIntegration.factory({ config }),
+      github: GitHubIntegration.factory({ config }),
+      gitlab: GitLabIntegration.factory({ config }),
+    });
+  }
+
+  constructor(integrationsByType: IntegrationsByType) {
+    this.byType = integrationsByType;
   }
 
-  constructor(private readonly integrations: ScmIntegrationPredicateTuple[]) {}
+  get azure(): ScmIntegrationsGroup<AzureIntegration> {
+    return this.byType.azure;
+  }
+
+  get bitbucket(): ScmIntegrationsGroup<BitbucketIntegration> {
+    return this.byType.bitbucket;
+  }
+
+  get github(): ScmIntegrationsGroup<GitHubIntegration> {
+    return this.byType.github;
+  }
+
+  get gitlab(): ScmIntegrationsGroup<GitLabIntegration> {
+    return this.byType.gitlab;
+  }
 
   list(): ScmIntegration[] {
-    return this.integrations.map(i => i.integration);
+    return Object.values(this.byType).flatMap(
+      i => i.list() as ScmIntegration[],
+    );
+  }
+
+  byUrl(url: string | URL): ScmIntegration | undefined {
+    return Object.values(this.byType)
+      .map(i => i.byUrl(url))
+      .find(Boolean);
   }
 
-  byUrl(url: string): ScmIntegration | undefined {
-    return this.integrations.find(i => i.predicate(new URL(url)))?.integration;
+  byHost(host: string): ScmIntegration | undefined {
```

Comment:
Kinda feeling that `list()` and `byUrl` should be enough? What are you thinking that this and the typed versions are useful for? They're pretty easy to add in the future if needed

Reply:
Note that the "generic" type returned by `ScmIntegrations.byUrl` at this point is largely useless. You can get a type id and a title, not the config or anything like that. Before we settle on what a generic interface for all integrations looks like, I wager that the type-aware consumers (like the scaffolder) want to use the typed integration groups.

Reply:
@freben Thinking that we main (and only?) use-case atm is to use `list()` to set something up that is able to interact with all the different integrations, like UrlReader, preparers, etc.

Reply:
I was rather thinking that for example [this code](https://github.com/backstage/backstage/blob/master/plugins/scaffolder-backend/src/scaffolder/stages/prepare/github.ts#L62) would want to do `this.integrations.github.byUrl(location).config.token`.

Yes, the frontend code might use `integrations.list()` to get the list of possible targets to select or something like that.